QgsPoint2 add project with 3D support

lbartoletti opened the pull request:
Since I work on an integration of different CAD tools, I suggest first a writing of project for QgsPointV2 supporting 3D.

lbartoletti committed:
QgsPoint2 add project with 3D support

nyalldawson commented on python/core/geometry/qgspointv2.sip:
```diff
@@ -179,6 +179,15 @@ class QgsPointV2: public QgsAbstractGeometry
+     * in a specified angle.
+     * @param distance distance to project
+     * @param azimuth angle to project in X Y, clockwise in degrees starting from north
+     * @param inclinaison angle to project in Z (3D)
```
Inclination

nyalldawson commented on src/core/geometry/qgspointv2.cpp:
```diff
@@ -462,3 +462,23 @@ double QgsPointV2::azimuth( const QgsPointV2& other ) const
+QgsPointV2 QgsPointV2::project( double distance, double azimuth, double inclinaison ) const
+{
+    double rads_xy = azimuth * M_PI / 180.0;
+    if ( !is3D() )
```
This looks unexpected to me. If I project a 2D point using an inclination I'd expect a 3D point result projected from an initial 0 z value.

lbartoletti replied:
I also thought of several ways to implement this method.

- distinguish a function for a 2D point (as a project2D method) and one for a 3D point (as a project3D method)
- consider that if we pass a 2D point then there is a projection only in 2D (the case for QgsPoint.project) and if we have a 3D point then we project in 3D (my proposal)
- consider that if we pass a point has necessarily a projection including in 3D (your expectation)

I did not propose the latter because it will force always to return a PointZ which for my few cases is not necessarily requested or desired. So you have to addZ for your QgsPointV2 (2D) before use project for have 3D projection.

nyalldawson replied:
What about:

- if inclination is 0 AND original point is 2D, return a 2D point
- else return 3D point

This would help maintain 2D points for those not interested in 3D, whilst keeping the API behaviour predictable.

Thinking about it, you'll also need to add m value handling here. Ie maintain any existing m value.

lbartoletti replied:
It's ok for me. Just inclination is 90 not 0.

nyalldawson reviewed (changes requested):
Looks like a handy method! Thanks!

lbartoletti committed:
Spell:
inclinaison -> inclination
correponds -> correspond

Algorithm:
Change requested from nyalldawson
Keep M value

nyalldawson commented on src/core/geometry/qgspointv2.cpp:
```diff
@@ -464,21 +464,33 @@ double QgsPointV2::azimuth( const QgsPointV2& other ) const
+
+    inclination = fmod( inclination, 360.0 );
+
+    if ( !is3D() && inclination == 90.0 )
```
I think it would be slightly safer to do:

    if ( !is3D() && qgsDoubleNear( inclination, 90.0 ) )

qgsDoubleNear allows for a tiny tolerance to account for floating point calculations.

I'm also wondering... what about an inclination of 270? Probably an extreme corner case, but worth covering too.

nyalldawson commented:
Thanks! I'm much happier with this behaviour.

nyalldawson commented:
Sorry - one last thing. Can you update the doxygen comments to reflect the logic wrt when a 2d or 3d point is returned?

lbartoletti committed:
* Add full doc with example for qgspointv2.h qgspointv2.sip.
* Use double compare function for inclination in qgspointv2.cpp.
* Change project from qgsexpression with qgspointv2.project.

nyalldawson commented:
Great stuff! thanks!
